src/libraries/Common/src/Roslyn/GetBestTypeByMetadataName.cs

```diff
@@ -134,5 +138,34 @@ private enum SymbolVisibility
             Private = 2,
             Friend = Internal,
         }
+
+        internal static bool HasAttributeSuffix(this string name, bool isCaseSensitive)
+        {
+            const string AttributeSuffix = "Attribute";
+
+            var comparison = isCaseSensitive ? StringComparison.Ordinal : StringComparison.OrdinalIgnoreCase;
+            return name.Length > AttributeSuffix.Length && name.EndsWith(AttributeSuffix, comparison);
+        }
+
+        public static ImmutableArray<T> ToImmutableArray<T>(this ReadOnlySpan<T> span)
+        {
+            if (span.Length == 0)
+                return ImmutableArray<T>.Empty;
+
+            var builder = ImmutableArray.CreateBuilder<T>(span.Length);
+            foreach (var item in span)
+                builder.Add(item);
+
+            return builder.MoveToImmutable();
+        }
```

Comment:
unfortunate, but there appears to be no non-allocating way to do this in the runtime currently.

Reply:
How important is it?  If it's important, you can do the inverse of
https://github.com/dotnet/runtime/blob/807d0916c5205d75ecb04809f8106a3caa46bbe4/src/libraries/System.Collections.Immutable/tests/ImmutableArrayTest.cs#L2615-L2638
with `span.ToArray()`.

Also, the builder itself is an allocation.  If a particular length is really common, like span.Length == 1, you could special-case it to just create an ImmutableArray from span[0] without going through a builder.

Reply:
> How important is it?

it's likely ok, just an unfortuanate wart.  i will optimize the 0/1 cases at least :)

Reply:
in Roslyn we use a pooled ImmutableArray.Builder, so that way teh allocation cost is ammortized :)

Reply:
> in Roslyn we use a pooled ImmutableArray.Builder, so that way teh allocation cost is ammortized :)

But it still allocates the array, yes?

Reply:
the array for ther ImmutableArray is allocated every time no matter what.  The pooled builder approach means that hte builder itself isn't garbage.

There's also the builder's internal array which can be pooled.  But that isn't an issue here since we know the exact lengths, and thus can use MoveToImmutable.

So really, it's just about saving the allocation for the builder obj. :)